src/imap.rs

```diff
@@ -1399,92 +1404,117 @@ impl Imap {
                 }
             };
 
-            let mut uid_msgs = server_uids
-                .iter()
-                .map(|&uid| (uid, None))
-                .collect::<HashMap<_, _>>();
-            let mut server_uids_it = server_uids.iter().peekable();
+            // Map from UIDs to unprocessed FETCH results. We put unprocessed FETCH results here
+            // when we want to process other messages first.
+            let mut uid_msgs = HashMap::with_capacity(request_uids.len());
+
             let mut count = 0;
-            while let Some(&&server_uid) = server_uids_it.peek() {
-                let mut msg = uid_msgs.insert(server_uid, None).flatten();
-                while msg.is_none() {
-                    let msg_unwrapped = match msgs.next().await {
-                        Some(Ok(msg)) => msg,
-                        Some(Err(_)) => continue,
-                        None => break,
+            for &request_uid in request_uids.iter() {
+                // Check if FETCH response is already in `uid_msgs`.
+                let mut fetch_response = uid_msgs.remove(&request_uid);
+
+                // Try to find a requsted UID in returned FETCH responses.
+                while fetch_response.is_none() {
+                    let next_fetch_response =
+                        if let Some(next_fetch_response) = fetch_responses.next().await {
+                            next_fetch_response
+                        } else {
+                            // No more FETCH responses received from the server.
+                            break;
+                        };
+
+                    let next_fetch_response = match next_fetch_response {
+                        Ok(next_fetch_response) => next_fetch_response,
+                        Err(err) => {
+                            warn!(context, "Failed to process IMAP FETCH result: {}.", err);
+                            continue;
+                        }
                     };
-                    let msg_uid = msg_unwrapped.uid.unwrap_or_default();
-                    if !uid_msgs.contains_key(&msg_uid) {
-                        // Unwanted UIDs are possible because of unsolicited responses, e.g. if
-                        // another client changes \Seen flag on a message after we do a prefetch but
-                        // before fetch. It's not an error if we receive such unsolicited response.
-                        continue;
-                    }
-                    msg = Some(msg_unwrapped);
-                    if msg_uid != server_uid && uid_msgs.insert(msg_uid, msg.take()).is_some() {
-                        warn!(context, "Got duplicated UID {}", msg_uid);
+
+                    if let Some(next_uid) = next_fetch_response.uid {
+                        if next_uid == request_uid {
+                            fetch_response = Some(next_fetch_response);
+                        } else if !request_uids.contains(&next_uid) {
```

Comment:
There's a bound actually on request_uids size, but i'd left a comment here so that readers of the code don't worry about O(n^2)